docs/mergeop/2_conflict_detection.md

```diff
@@ -0,0 +1,205 @@
+
+# **Appendix**
+
+## Future Merge Options
+
+While they won’t be included in the initial implementation, eventually additional options such as the following should be possible to add-on:
+
+1. DiscardDeletions bool
+    1. Enables/disables whether whiteout/opaques will be completely discarded and not have any effect on the merge result
+        1. This is useful for emulation of Dockerfile’s COPY
+    2. For the efficient overlay implementation, this can be accomplished through use of nested overlays (as once an overlay is created, whiteouts+opaques are no longer visible, making them hidden when the overlay mount is used as a lowerdir)
+2. ConflictHandlers []ConflictHandler
+    1. Allows file conflicts to be resolved by arbitrary code (maybe an ExecOp that will be setup with mounts that provide the conflicting files and a mount to write the resolution under)
+
+## Conflict Detection Interface
+
+This interface will be internal-only, but will look something like this:
+
+```
+import "io/fs"
+
+type FSMetadata interface {
+    Stat(name string) (fs.FileInfo, error)
+    ReadDir(name string) ([]fs.DirEntry, error)
+    // If this FSMetadata was created by merging other FSMetadatas, then
+    // Lowers will return those others making up the merge
+    Lowers() []FSMetadata
+
+    // Conflicts will return any conflicts that are present due to the merging
+    // of this FSMetadata’s Lowers (if any). Importantly, it does not 
+    // "recurse" in that if there is a lower that itself is a merge w/
+    // conflicts, those will not also be returned here. This is important
+    // because we may, for example, often import images that have multiple
+    // layers with conflicts, but we don’t care about those in this "level"
+    // of the merge.
+    Conflicts() []Conflict
+}
+
+type ConflictType int
+
+const (
+    Whiteout ConflictType = iota
+    Opaque
+    FileOverwrite
+    DirectoryOverwrite
+)
+
+type Conflict struct {
+    Type ConflictType
+    Path string
+    // higher is the merged FSMetadata that overrides in this conflict
+    Higher FSMetadata
+    // lower is the merged FSMetadata that gets overridden in this conflict
+    Lower FSMetadata
+}
+
+func Merge(lowers []FSMetadata) FSMetadata {
+}
+
+// ReadLayer is similar to the Differ in that it compares two directories, but
+// in this case it returns an FSMetadata object instead of writing a tarball
+// to the content store. This will be used when we need to first load the
+// metadata for a snapshot.
+
+func ReadLayer(lowerdir, upperdir string) FSMetadata {
+    // Implementation will use continuity’s diff utils, but with a changeFunc
+    // that creates an FSMetadata
+}
+
+func Marshal(fs FSMetadata) ([]byte, error) {
+}
+
+func Unmarshal(marshalled []byte) (FSMetadata, error) {
+}
+```
+
+## Conflict Detection Alternatives
+
+There are a few existing userspace emulations of overlayfs that were considered for implementing our layer conflict detection features:
```

Comment:
Why do we need userspace filesystem for conflict calculation? Isn't it enough to just compare filesystem metadata between layers? (maybe I'm missing something)

Reply:
No worries, yeah these are the alternatives, I just wanted to mention them to give context on why we're proposing doing our own custom code instead of using a library.

Added an extra sentence clarifying this.